Tasks may fail on reconnect if alloc and task runners run before client connects to servers

### Nomad version
`Nomad v1.4.1 (2aa7e66bdb526e25f59883952d74dad7ea9a014e)`

### Operating system and Environment details
macOs

### Issue

Jobs that need alloc or task runner hooks that communicate with Nomad servers (such as Nomad Native Service Discovery) may fail on reconnect if these hooks run before the client establishes a connection with the server.

<img width="1239" alt="image" src="https://user-images.githubusercontent.com/775380/199851840-b13ac90a-3ef4-4aea-9f9e-0e7b37c29001.png">

### Reproduction steps

1. Start a 2 client Nomad cluster
2. Run job
3. Make sure each client has one allocation and stop one of the clients
4. Wait for the allocation to be `unknown`
5. Restart client

This is a race condition problem, so you may need to try this a few times 😅 

#### Expected Result

Allocation reconnects properly.

#### Actual Result

Allocation fails on reconnect.

### Job file (if appropriate)

```hcl
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    count = 2
    max_client_disconnect = "5m"

    network {
      port "db" {
        to = 6379
      }
    }

    service {
      name     = "redis"
      port     = "db"
      provider = "nomad"
    }

    task "redis" {
      driver = "docker"

      config {
        image          = "redis:6"
        ports          = ["db"]
        auth_soft_fail = true
      }
    }
  }
}
```

### Sample configuration files

```hcl
name      = "server1"
data_dir  = "/tmp/nomad/server"
log_level = "TRACE"

ports {
  http = 4646
  rpc  = 4647
  serf = 4648
}

server {
  enabled           = true
  bootstrap_expect  = 1
  heartbeat_grace   = "1s"
  min_heartbeat_ttl = "5s"
}

client {
  enabled = true
}
```

```hcl
name      = "client1"
data_dir  = "/tmp/nomad/client1"
log_level = "TRACE"

ports {
  http = 4656
  rpc  = 4657
  serf = 4658
}

server {
  enabled = false
}

client {
  enabled = true

  server_join {
    retry_join = ["127.0.0.1"]
  }
}
```

### Nomad Client logs (if appropriate)

```
    2022-11-03T19:03:37.363-0400 [DEBUG] client: registration waiting on servers
    2022-11-03T19:03:37.363-0400 [ERROR] client: error discovering nomad servers: error="client.consul: unable to query Consul datacenters: Get \"http://127.0.0.1:8500/v1/catalog/datacenters\": dial tcp 127.0.0.1:8500: connect: connection refused"
    2022-11-03T19:03:37.364-0400 [TRACE] client.alloc_runner.task_coordinator: state transition: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd from=init to=init
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner.task_coordinator: state transition: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd from=init to=prestart
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner.task_coordinator: state transition: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd from=prestart to=main
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner.task_coordinator: state transition: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd from=main to=poststart
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner.task_coordinator: state transition: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd from=poststart to=wait_alloc
    2022-11-03T19:03:37.372-0400 [INFO]  client: started client: node_id=26683602-1db8-25b6-e1d8-169a28aac063
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hooks: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd start="2022-11-03 19:03:37.372683 -0400 EDT m=+6.217725751"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=alloc_dir start="2022-11-03 19:03:37.372707 -0400 EDT m=+6.217749460"
    2022-11-03T19:03:37.372-0400 [TRACE] consul.sync: commit sync operations: ops="<1, 1, 0, 0>"
    2022-11-03T19:03:37.372-0400 [WARN]  client.server_mgr: no servers available
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=alloc_dir end="2022-11-03 19:03:37.37292 -0400 EDT m=+6.217962335" duration="212.875µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=cgroup start="2022-11-03 19:03:37.372925 -0400 EDT m=+6.217967876"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=cgroup end="2022-11-03 19:03:37.372928 -0400 EDT m=+6.217970168" duration="2.292µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=await_previous_allocations start="2022-11-03 19:03:37.37293 -0400 EDT m=+6.217972251"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=await_previous_allocations end="2022-11-03 19:03:37.372932 -0400 EDT m=+6.217974126" duration="1.875µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=migrate_disk start="2022-11-03 19:03:37.372938 -0400 EDT m=+6.217980085"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=migrate_disk end="2022-11-03 19:03:37.372943 -0400 EDT m=+6.217985168" duration="5.083µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=alloc_health_watcher start="2022-11-03 19:03:37.372945 -0400 EDT m=+6.217987543"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner.runner_hook.alloc_health_watcher: not watching; already has health set: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=alloc_health_watcher end="2022-11-03 19:03:37.372948 -0400 EDT m=+6.217990543" duration="3µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=network start="2022-11-03 19:03:37.37295 -0400 EDT m=+6.217992835"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=network end="2022-11-03 19:03:37.372952 -0400 EDT m=+6.217994418" duration="1.583µs"
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: running pre-run hook: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd name=group_services start="2022-11-03 19:03:37.372954 -0400 EDT m=+6.217996710"
    2022-11-03T19:03:37.372-0400 [DEBUG] http: UI is enabled
    2022-11-03T19:03:37.372-0400 [WARN]  client.server_mgr: no servers available
    2022-11-03T19:03:37.372-0400 [TRACE] client.alloc_runner: finished pre-run hooks: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd end="2022-11-03 19:03:37.372995 -0400 EDT m=+6.218037835" duration="312.084µs"
    2022-11-03T19:03:37.372-0400 [ERROR] client.alloc_runner: prerun failed: alloc_id=61d26e2f-3f16-5149-614a-c23f1c0199fd error="pre-run hook \"group_services\" failed: no servers"
    2022-11-03T19:03:37.373-0400 [DEBUG] http: UI is enabled
    2022-11-03T19:03:37.373-0400 [INFO]  agent.joiner: starting retry join: servers=127.0.0.1
    2022-11-03T19:03:37.374-0400 [DEBUG] client.server_mgr: new server list: new_servers=[127.0.0.1:4647] old_servers=[]
    2022-11-03T19:03:37.374-0400 [INFO]  agent.joiner: retry join completed: initial_servers=1 agent_mode=client
```

--- a/.changelog/15140.txt
+++ b/.changelog/15140.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+client: prevent allocations from failing on client reconnect by retrying RPC requests when no servers are available yet
+```

--- a/client/rpc.go
+++ b/client/rpc.go
@@ -109,7 +109,7 @@ TRY:
 			info.SetTimeToBlock(0)
 			return c.RPC(method, args, reply)
 		}
-		c.rpcLogger.Error("error performing RPC to server, deadline exceeded, cannot retry", "error", rpcErr, "rpc", method, "server", server.Addr)
+		c.rpcLogger.Error("error performing RPC to server, deadline exceeded, cannot retry", "error", rpcErr, "rpc", method)
 		return rpcErr
 	}
 

--- a/client/rpc.go
+++ b/client/rpc.go
@@ -75,9 +75,7 @@ TRY:
 	server := c.servers.FindServer()
 	if server == nil {
 		rpcErr = noServersErr
-	}
-
-	if server != nil {
+	} else {
 		// Make the request.
 		rpcErr = c.connPool.RPC(c.Region(), server.Addr, method, args, reply)
 

--- a/client/rpc.go
+++ b/client/rpc.go
@@ -70,34 +70,39 @@ func (c *Client) RPC(method string, args interface{}, reply interface{}) error {
 	}
 
 TRY:
+	var rpcErr error
+
 	server := c.servers.FindServer()
 	if server == nil {
-		return noServersErr
+		rpcErr = noServersErr
 	}
 
-	// Make the request.
-	rpcErr := c.connPool.RPC(c.Region(), server.Addr, method, args, reply)
+	if server != nil {
+		// Make the request.
+		rpcErr = c.connPool.RPC(c.Region(), server.Addr, method, args, reply)
 
-	if rpcErr == nil {
-		c.fireRpcRetryWatcher()
-		return nil
+		if rpcErr == nil {
+			c.fireRpcRetryWatcher()
+			return nil
+		}
+
+		// If shutting down, exit without logging the error
+		select {
+		case <-c.shutdownCh:
+			return nil
+		default:
+		}
+
+		// Move off to another server, and see if we can retry.
+		c.rpcLogger.Error("error performing RPC to server", "error", rpcErr, "rpc", method, "server", server.Addr)
+		c.servers.NotifyFailedServer(server)
+
+		if !canRetry(args, rpcErr) {
+			c.rpcLogger.Error("error performing RPC to server which is not safe to automatically retry", "error", rpcErr, "rpc", method, "server", server.Addr)
+			return rpcErr
+		}
 	}
 
-	// If shutting down, exit without logging the error
-	select {
-	case <-c.shutdownCh:
-		return nil
-	default:
-	}
-
-	// Move off to another server, and see if we can retry.
-	c.rpcLogger.Error("error performing RPC to server", "error", rpcErr, "rpc", method, "server", server.Addr)
-	c.servers.NotifyFailedServer(server)
-
-	if !canRetry(args, rpcErr) {
-		c.rpcLogger.Error("error performing RPC to server which is not safe to automatically retry", "error", rpcErr, "rpc", method, "server", server.Addr)
-		return rpcErr
-	}
 	if time.Now().After(deadline) {
 		// Blocking queries are tricky.  jitters and rpcholdtimes in multiple places can result in our server call taking longer than we wanted it to. For example:
 		// a block time of 5s may easily turn into the server blocking for 10s since it applies its own RPCHoldTime. If the server dies at t=7s we still want to retry

